OpenRA.Mods.Common/Traits/DockManager.cs

```diff
@@ -0,0 +1,443 @@
+﻿#region Copyright & License Information
+/*
+ * Copyright 2007-2017 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using OpenRA.Activities;
+using OpenRA.Mods.Common.Activities;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	public class DockManagerInfo : ITraitInfo, Requires<DockInfo>
+	{
+		[Desc("Are any of the docks lie outside the building footprint? (and needs obstacle checking)")]
+		public readonly bool ExternalDocks = false;
+
+		[Desc("Dock next to the actor like RA1 naval yard?",
+			"Although dock position is ignored, one dummy dock is still required to determine DockAngle and stuff when docking.")]
+		public readonly bool DockNextToActor = false;
+
+		[Desc("Enable deadlock detection")]
+		public readonly bool DeadlockDetectionEnabled = true;
+
+		[Desc("Dead lock detection sampling is done this often.")]
+		public readonly int DeadlockDetectionPeriod = 457; // prime number yay =~ 30 seconds
+
+		public object Create(ActorInitializer init) { return new DockManager(init, this); }
+	}
+
+	/*
+	 * The class to do all the crazy queue management.
+	 * Not all multi-dock guys need queue management so making a separate manager class.
+	 */
+	public class DockManager : ITick, INotifyKilled, INotifyActorDisposing
+	{
+		readonly DockManagerInfo info;
+		readonly Actor host; // Don't use "self" as the name. Eventually, it gets confusing with client and causes bug!
```

Comment:
Can you elaborate on where this causes a problem? `self` always refers to the actor who owns the trait.  If methods are being called in ways where this becomes confusing, then that suggests deeper problems.

Reply:
It wasn't anything deep, it gets confusing if you keep switching the file you are working on, as there are many files involved, including interfaces, activities, and DockClient. At this stage, the files seem quite stable and I think I can rename this back to self.